### PR TITLE
[VAULT-1828]Decouple read and handling in client to proxy connection

### PR DESCRIPTION
**Pull request by yuriyshafranyuk1**
add new property `separateProcessingEventLoop` to proxy server configuration that allow to read and handle request in separate threads

### CODE REVIEW

**Review by viacheslav-rud — APPROVED** [review]

File: src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java [hunk 1]
```diff
@@ -840,8 +855,29 @@ private void initChannelPipeline(ChannelPipeline pipeline) {
...
+        } else {
+            pipeline.addLast("handler", this);
+        }
+
+    }
 
+    Runnable wrapTask(Runnable task) {
+        return () -> {
+            final Optional<GlobalStateHandler> globalStateHandler =
+                Optional.ofNullable(proxyServer.getGlobalStateHandler());
```
Comment (viacheslav-rud) [inline src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java]: I don't see a benefit of using `Optional` here over simple if-statement

**Review by viacheslav-rud — COMMENTED** [review]
Aren't we gonna cover these changes with tests somehow?

File: src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java [hunk 2]
```diff
@@ -840,8 +854,32 @@ private void initChannelPipeline(ChannelPipeline pipeline) {
             pipeline.addLast("outboundGlobalStateHandler", new OutboundGlobalStateHandler(this));
         }
 
-        pipeline.addLast("handler", this);
+        if(separateProcessingEventLoop) {
```
Comment (yuriyshafranyuk1) [inline src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java:851]: allow to run main handler of client to proxy connection in separate thread. So actually it allow to decouple read and request handling.

File: docker-compose.yaml [hunk 3]
```diff
@@ -5,6 +5,7 @@ services:
     image: maven:3.8.6-jdk-8-slim
     command: bash -c "apt-get update &&
       apt-get install git -y &&
+      git config --global --add safe.directory '*' &&
```
Comment (Iapetus999) [inline docker-compose.yaml]: Probably can drop these lines related to git, don't think we need them

File: docker-compose.yaml [hunk 4]
```diff
@@ -5,8 +5,6 @@ services:
     image: maven:3.8.6-jdk-8-slim
     command: bash -c "apt-get update &&
       apt-get install git -y &&
```
Comment (Iapetus999) [inline docker-compose.yaml:7]: prob can kill this line too

File: src/main/java/org/littleshoot/proxy/HttpProxyServerBootstrap.java [hunk 5]
```diff
@@ -460,4 +460,7 @@ HttpProxyServerBootstrap withConnectTimeout(
      * @param acceptorLoggingEnabled
      */
     HttpProxyServerBootstrap withAcceptorLoggingEnabled(boolean acceptorLoggingEnabled);
+
+
+    HttpProxyServerBootstrap withSeparateProcessingEventLoop(boolean separateProcessingEventLoop);
```
Comment (Iapetus999) [inline src/main/java/org/littleshoot/proxy/HttpProxyServerBootstrap.java:477]: add javadoc

File: src/main/java/org/littleshoot/proxy/RequestTracer.java [hunk 6]
```diff
@@ -8,11 +8,14 @@ public interface RequestTracer {
    * Start tracing proxy request
    * @param channel
    */
-  void start(Channel channel);
+  void start(Channel channel, Object msg);
```
Comment (Iapetus999) [inline src/main/java/org/littleshoot/proxy/RequestTracer.java:14]: Is there a way to type or template msg?

Comment (Iapetus999) [reply src/main/java/org/littleshoot/proxy/RequestTracer.java:14]: Also document `msg`

File: src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java [hunk 7]
```diff
@@ -840,8 +854,32 @@ private void initChannelPipeline(ChannelPipeline pipeline) {
...
+            EventLoop eventExecutor = this.proxyServer.getServerGroup()
+                .getClientToProxyProcessingPoolForTransport(proxyServer.getTransportProtocol()).next();
+
+            EventExecutor globalStateWrapperEvenLoop = new GlobalStateWrapperEventLoop(this, eventExecutor.next());
```
Comment (Iapetus999) [inline src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java]: typo `event`

File: src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java [hunk 8]
```diff
@@ -1445,7 +1483,10 @@ public InetSocketAddress getClientAddress() {
         if (channel == null) {
             return null;
         }
-        return (InetSocketAddress) channel.remoteAddress();
+        if(channel.remoteAddress() instanceof InetSocketAddress) {
+            return (InetSocketAddress) channel.remoteAddress();
+        }
+        return null;
```
Comment (Iapetus999) [inline src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java:1467]: What about `0.0.0.0` for null

File: src/main/java/org/littleshoot/proxy/impl/GlobalStateWrapperEventLoop.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+import io.netty.util.concurrent.Promise;
+import io.netty.util.concurrent.ScheduledFuture;
+
+public class GlobalStateWrapperEventLoop implements EventExecutor {
+
+  private final ClientToProxyConnection connection;
+
+  private final EventExecutor eventLoop;
+
+  GlobalStateWrapperEventLoop(ClientToProxyConnection connection) {
```
Comment (Iapetus999) [inline src/main/java/org/littleshoot/proxy/impl/GlobalStateWrapperEventLoop.java]: `public`?

File: src/main/java/org/littleshoot/proxy/impl/GlobalStateWrapperEventLoop.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+import java.util.concurrent.TimeoutException;
+
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.concurrent.EventExecutorGroup;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.ProgressivePromise;
+import io.netty.util.concurrent.Promise;
+import io.netty.util.concurrent.ScheduledFuture;
+
+public class GlobalStateWrapperEventLoop implements EventExecutor {
```
Comment (Iapetus999) [inline src/main/java/org/littleshoot/proxy/impl/GlobalStateWrapperEventLoop.java:21]: javadoc saying this is facade for event loop

File: src/main/java/org/littleshoot/proxy/impl/ProxyConnection.java [hunk 11]
```diff
@@ -762,7 +762,7 @@ protected class RequestTracerHandler extends ChannelDuplexHandler {
         @Override
         public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
             try {
-                proxyServer.getRequestTracer().start(clientToProxyConnection.channel);
+                proxyServer.getRequestTracer().start(clientToProxyConnection.channel, msg);
```
Comment (Iapetus999) [inline src/main/java/org/littleshoot/proxy/impl/ProxyConnection.java:765]: curious to know if there's any way to type msg, ok if we can't 😒

File: src/main/java/org/littleshoot/proxy/impl/ReadLoggingHandler.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    super(LogLevel.DEBUG);
+    this.requestTracer = requestTracer;
+  }
+
+  @Override
+  public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress, SocketAddress localAddress,
+      ChannelPromise promise) throws Exception {
+    this.logger.log(InternalLogLevel.INFO, this.format(ctx, "Connected"));
+    ctx.connect(remoteAddress, localAddress, promise);
+  }
```
Comment (Iapetus999) [inline src/main/java/org/littleshoot/proxy/impl/ReadLoggingHandler.java:25]: Thought: Really wish we could integrate vgs-logging here so it dumps the standard fields etc

File: src/test/java/org/littleshoot/proxy/impl/SeparateProcessingThreadTest.java [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+              public io.netty.handler.codec.http.HttpResponse clientToProxyRequest(HttpObject httpObject) {
+                if (httpObject instanceof FullHttpRequest &&
+                    Thread.currentThread().getName().contains("ClientToProxyProcessorWorker")) {
+                  try {
+                    String[] uriSplit = ((FullHttpRequest) httpObject).uri().split("/");
+                    String key = this.ctx.channel().attr(traceKey).get();
+                    Assert.assertEquals(uriSplit[uriSplit.length - 1], key);
+                    traces.get(key).incrementAndGet();
+                  } catch (Exception e) {
+                    e.printStackTrace();
```
Comment (Iapetus999) [inline src/test/java/org/littleshoot/proxy/impl/SeparateProcessingThreadTest.java]: not fail test?

File: src/test/java/org/littleshoot/proxy/impl/SeparateProcessingThreadTest.java [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+                    withHeader("key", key)
+                    .withBody("{\"status\":\"OK\"}"));
+            Thread.sleep(100);
+            org.apache.http.HttpResponse response = HttpClientUtil.performHttpPost(
+                "http://localhost:" + mockServerPort + "/traceRequest/" + key, 100, proxyServer);
+            Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+            traces.get(key).incrementAndGet();
+            countDownLatch.countDown();
+          } catch (Exception e) {
+            e.printStackTrace();
```
Comment (Iapetus999) [inline src/test/java/org/littleshoot/proxy/impl/SeparateProcessingThreadTest.java]: fail test?

File: src/test/java/org/littleshoot/proxy/impl/SeparateProcessingThreadTest.java [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+          }
+        }
+      });
+    }
+    countDownLatch.await(5, TimeUnit.SECONDS);
```
Comment (Iapetus999) [inline src/test/java/org/littleshoot/proxy/impl/SeparateProcessingThreadTest.java]: You might be able to capture the logging and verify the trace IDs

File: src/main/java/org/littleshoot/proxy/impl/ReadLoggingHandler.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+  @Override
+  public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+    this.logger.log(InternalLogLevel.INFO, this.format(ctx,
+        this.format(ctx, "Read message. trace id " + requestTracer.getCurrentTrace(ctx.channel()))));
+    ctx.fireChannelRead(msg);
+  }
+
+  @Override
+  public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+    this.logger.log(InternalLogLevel.INFO,
```
Comment (Iapetus999) [inline src/main/java/org/littleshoot/proxy/impl/ReadLoggingHandler.java:36]: Needs to be a test for this code

File: src/test/java/org/littleshoot/proxy/impl/SeparateProcessingThreadTest.java [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+    for (AtomicInteger atomicInteger : traces.values()) {
+      Assert.assertEquals(4, atomicInteger.get());
+    }
+    Assert.assertEquals(requestsAmount, traces.size());
+  }
```
Comment (Iapetus999) [inline src/test/java/org/littleshoot/proxy/impl/SeparateProcessingThreadTest.java:269]: If you don't use the new event loop, will this test fail?

Comment (Iapetus999) [reply src/test/java/org/littleshoot/proxy/impl/SeparateProcessingThreadTest.java:269]: `separateProcessingEventLoop` = false

File: docker-compose.yaml (hunk 4 above)
Comment (yuriyshafranyuk1) [reply docker-compose.yaml:7]: mvn fails without git

File: src/main/java/org/littleshoot/proxy/impl/GlobalStateWrapperEventLoop.java (hunk 10 above)
Comment (yuriyshafranyuk1) [reply src/main/java/org/littleshoot/proxy/impl/GlobalStateWrapperEventLoop.java:21]: done

**Review by Iapetus999 — APPROVED** [review]

**Pull request merged**